crates/sui-rpc-benchmark/src/direct/query_executor.rs

```diff
@@ -0,0 +1,201 @@
+        let raw_rows = rows.iter().map(Self::row_to_values).collect();
+
+        Ok(EnrichedBenchmarkQuery {
+            query: query.clone(),
+            rows: raw_rows,
+            types,
+        })
+    }
+
+    pub async fn initialize_samples(&mut self) -> Result<()> {
+        for query in &self.queries.clone() {
+            let enriched = self.enrich_query(query).await?;
+            self.enriched_queries.push(enriched);
+        }
+        Ok(())
+    }
+
+    async fn worker_task(
+        pool: Pool<PostgresConnectionManager<NoTls>>,
+        enriched_queries: Vec<EnrichedBenchmarkQuery>,
+        metrics: MetricsCollector,
+        deadline: Instant,
+    ) -> Result<()> {
+        let client = pool.get().await?;
+        while Instant::now() < deadline {
+            let enriched = enriched_queries
+                .choose(&mut rand::thread_rng())
+                .ok_or_else(|| anyhow::anyhow!("No queries available"))?;
+
+            let row = match enriched.rows.choose(&mut rand::thread_rng()) {
+                Some(row) => row,
+                None => {
+                    // skip when the table is empty and thus no values to sample.
+                    continue;
+                }
+            };
+
+            let params: Vec<Box<dyn ToSql + Sync + Send>> = row
+                .iter()
+                .map(|val| match val {
+                    SqlValue::Text(v) => Box::new(v) as Box<dyn ToSql + Sync + Send>,
+                    SqlValue::Int4(v) => Box::new(v) as Box<dyn ToSql + Sync + Send>,
+                    SqlValue::Int8(v) => Box::new(v) as Box<dyn ToSql + Sync + Send>,
+                    SqlValue::Float8(v) => Box::new(v) as Box<dyn ToSql + Sync + Send>,
+                    SqlValue::Bool(v) => Box::new(v) as Box<dyn ToSql + Sync + Send>,
+                    SqlValue::Int2(v) => Box::new(v) as Box<dyn ToSql + Sync + Send>,
+                    SqlValue::Bytea(v) => Box::new(v) as Box<dyn ToSql + Sync + Send>,
+                })
+                .collect();
+            let param_refs: Vec<&(dyn ToSql + Sync)> = params
+                .iter()
+                .map(|p| p.as_ref() as &(dyn ToSql + Sync))
+                .collect();
+
+            let query_str = enriched.query.query_template.clone();
+
+            let start = Instant::now();
+            let result = client.query(&query_str, &param_refs[..]).await;
+
+            metrics.record_query(&enriched.query.table_name, start.elapsed(), result.is_err());
+        }
+        Ok(())
+    }
+
+    pub async fn run(&mut self) -> Result<BenchmarkResult> {
+        if self.enriched_queries.is_empty() {
+            self.initialize_samples().await?;
+        }
+
+        println!(
+            "Running benchmark with {} concurrent clients",
+            self.config.concurrency
+        );
+
+        let start = Instant::now();
+        let deadline = start + self.config.duration;
+
+        let queries_per_worker = self.enriched_queries.chunks(
```

Comment:
You could use `for_each_spawned` for this to save having to re-implement a worker pool -- it will also do a better job of balancing work between workers if tasks take different amounts of time to complete (and one worker's task queue drains faster than the rest).